### PR TITLE
added custom return object

### PR DESCRIPTION
**Pull request by puzzleshark**
Restores behavior of pennylane returning custom items to as it was here https://github.com/PennyLaneAI/pennylane/issues/1109#issuecomment-785616046

In this pr, the return object is still an autograd numpy array (which we require for differentiation) with a single entry (being the custom object) and so requires `.item()` to unpack (see the discussion linked above).

### CODE REVIEW

**Comment by cvjjm** [comment]
When running this test with the latest master merged in I get an additional error:
```
        with qml.tape.Unwrap(*tapes):
            res, jacs = execute_fn(tapes, **gradient_kwargs)

        for i, r in enumerate(res):

            if isinstance(r, np.ndarray):
                # For backwards compatibility, we flatten ragged tape outputs
                # when there is no sampling
                try:
>                   if isinstance(r[0][0], dict):
E                   TypeError: 'SpecialObject' object is not subscriptable

pennylane/interfaces/autograd.py:118: TypeError
```
To fix this `TypeError` probably needs to be added to the try block in autograd.py

File: pennylane/interfaces/autograd.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def _execute(
                 if isinstance(r[0][0], dict):
                     # This happens when measurement type is Counts and shot vector is passed
                     continue
-            except (IndexError, KeyError):
+            except (TypeError, IndexError, KeyError):
```
Comment (antalszava) [inline pennylane/interfaces/autograd.py:121]: What's the reason for this change?

File: tests/devices/test_default_qubit_autograd.py [hunk 2]
```diff
@@ -541,3 +543,120 @@ def test_partial_subsystem(self, mocker):
 
         assert np.all(res == state)
         spy.assert_called()
+
+
+@pytest.mark.autograd
+class TestObservableWithObjectReturnType:
```
Comment (antalszava) [inline tests/devices/test_default_qubit_autograd.py:549]: Likely this test case would better be fitted in the `tests/interfaces/test_autograd_qnode.py` file or `tests/test_qubit_device.py`. The feature is not specific to `default.qubit.autograd` (this test file is though).

File: tests/devices/test_default_qubit_autograd.py [hunk 3]
```diff
@@ -541,3 +543,120 @@ def test_partial_subsystem(self, mocker):
...
+        # PennyLane swaps out dev for default.qubit.autograd
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def qnode(x):
+            qml.RY(x, wires=0)
+            return qml.expval(SpecialObservable(wires=0))
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def reference_qnode(x):
+            qml.RY(x, wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
```
Comment (antalszava) [inline tests/devices/test_default_qubit_autograd.py:640]: Could we substitute this part with providing the analytic expression for the jacobian? 

I think we should have the value in our test suite somewhere and it would:
1. Not rely on the logic of PennyLane to check something;
2. Not require extra QNode evaluations.

File: tests/devices/test_default_qubit_autograd.py [hunk 4]
```diff
@@ -541,3 +543,120 @@ def test_partial_subsystem(self, mocker):
...
+        assert isinstance(out.item(), SpecialObject)
+        assert np.isclose(out.item().val, reference_qnode(0.2))
+        reference_jac = (qml.jacobian(reference_qnode)(np.array(0.2, requires_grad=True)),)
+
+        assert np.isclose(
+            reference_jac,
+            qml.jacobian(qnode)(np.array(0.2, requires_grad=True)).item().val,
+        )
+
+        # now check that also the device jacobian works with a custom return type
```
Comment (antalszava) [inline tests/devices/test_default_qubit_autograd.py:653]: This looks like a separate test case.

**Review by antalszava — CHANGES_REQUESTED** [review]
Looking good overall @puzzleshark, thank you for looking into this! :100: 

The main change to do from my side would be moving the test cases into another more relevant file.

File: tests/devices/test_default_qubit_autograd.py [hunk 5]
```diff
@@ -541,3 +543,120 @@ def test_partial_subsystem(self, mocker):
...
+                # jacobian for us but return it as a device jacobian
+                gradient_tapes, fn = qml.gradients.param_shift(tape)
+                tape_jacobian = fn(qml.execute(gradient_tapes, self, None))
+                return tape_jacobian
+
+        dev = DeviceSupportingSpecialObservable(wires=1, shots=None)
+
+        # force diff_method='parameter-shift' because otherwise
+        # PennyLane swaps out dev for default.qubit.autograd
+        @qml.qnode(dev, diff_method="parameter-shift")
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply tests/devices/test_default_qubit_autograd.py]: I think in theory it would but the "special Object" class does not have the necessary dunder methods for that

File: tests/devices/test_default_qubit_autograd.py (hunk 3 above)
Comment (puzzleshark) [reply tests/devices/test_default_qubit_autograd.py]: hmm, I think it's clearer as it is in showing that it is "equivalent" to the sample pl circuit.

File: tests/devices/test_default_qubit_autograd.py (hunk 4 above)
Comment (puzzleshark) [reply tests/devices/test_default_qubit_autograd.py]: separated out

File: tests/devices/test_default_qubit_autograd.py (hunk 2 above)
Comment (puzzleshark) [reply tests/devices/test_default_qubit_autograd.py]: fixed

File: pennylane/interfaces/autograd.py (hunk 1 above)
Comment (puzzleshark) [reply pennylane/interfaces/autograd.py:121]: > When running this test with the latest master merged in I get an additional error:
> 
> ```
>         with qml.tape.Unwrap(*tapes):
>             res, jacs = execute_fn(tapes, **gradient_kwargs)
> 
>         for i, r in enumerate(res):
> 
>             if isinstance(r, np.ndarray):
>                 # For backwards compatibility, we flatten ragged tape outputs
>                 # when there is no sampling
>                 try:
> >                   if isinstance(r[0][0], dict):
> E                   TypeError: 'SpecialObject' object is not subscriptable
> 
> pennylane/interfaces/autograd.py:118: TypeError
> ```
> 
> To fix this `TypeError` probably needs to be added to the try block in autograd.py

**Review by antalszava — APPROVED** [review]
Looks good! :tada: Great one :slightly_smiling_face:
